app/reports/invitation_uptake.rb

```diff
@@ -0,0 +1,35 @@
+class InvitationUptake
+  def run
+    CSV.generate do |csv|
+      csv << ['FCA Number', 'Registered Name', 'Invitation Sent', 'Invitation Accepted', 'Password Reset Sent']
+      query.all.each do |row|
+        csv << [
+          row.fca_number,
+          row.registered_name,
+          nice_date(row.invitation_sent_at),
+          nice_date(row.invitation_accepted_at),
+          nice_date(row.reset_password_sent_at)
+        ]
+      end
+    end
+  end
+
+  private
+
+  def query
+    User.where.not(invitation_sent_at: nil)
+      .joins(principal: [:firm])
+      .select(
+        'firms.fca_number',
+        'firms.registered_name',
+        'users.invitation_sent_at',
+        'users.invitation_accepted_at',
+        'users.reset_password_sent_at'
```

Comment:
I don't understand what we can assume about the `reset_password_sent_at` field for a user at this point. Maybe worth asking @JiggyPete. 

However, I think there are some other fields we can use to reason about how far they've made it through the invitation process:

1. We could show whether the `password` field is blank or not? I don't believe there is any value set in the password field until they've actually accepted the invite and set one. So this would be a good indicator that they've followed the process to that point.
2. We can find out if they got as far as actually signing in. There are `User.sign_in_count` and `User.last_sign_in_at` fields which would show that they've actually successfully logged in. I think it would be worth putting some test data into the system so we can follow an invite ourselves and see what fields change as we follow the process through.

What I'm expecting to get out should we total up the different types of states here is a funnel for example:

Invitations sent: 10
Invitations accepted: 9
Passwords set: 5
Successful Login least once: 4
Return visits already: 3